File: pylox/pylox/Stmt.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Final, Optional, override

from Token import Token
from Expr import Expr, Variable


class Stmt(ABC):
    @abstractmethod
    def accept[R](self, visitor: Visitor[R]) -> R: ...


class Visitor[R]:
    @abstractmethod
    def visit_Block_Stmt(self, stmt: Block) -> R: ...

    @abstractmethod
    def visit_Class_Stmt(self, stmt: Class) -> R: ...

    @abstractmethod
    def visit_Expression_Stmt(self, stmt: Expression) -> R: ...

    @abstractmethod
    def visit_Function_Stmt(self, stmt: Function) -> R: ...

    @abstractmethod
    def visit_If_Stmt(self, stmt: If) -> R: ...

    @abstractmethod
    def visit_Print_Stmt(self, stmt: Print) -> R: ...

    @abstractmethod
    def visit_Return_Stmt(self, stmt: Return) -> R: ...

    @abstractmethod
    def visit_Var_Stmt(self, stmt: Var) -> R: ...

    @abstractmethod
    def visit_While_Stmt(self, stmt: While) -> R: ...


class Block(Stmt):
    def __init__(self, statements: list[Stmt]):
        super().__init__()
        self.statements: Final[list[Stmt]] = statements

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Block_Stmt(self)


class Class(Stmt):
    def __init__(
        self, name: Token, super_class: Optional[Variable], methods: list[Function]
    ):
        super().__init__()
        self.name: Final[Token] = name
        self.super_class: Final[Optional[Variable]] = super_class
        self.methods: Final[list[Function]] = methods

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Class_Stmt(self)


class Expression(Stmt):
    def __init__(self, expression: Expr):
        super().__init__()
        self.expression: Final[Expr] = expression

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Expression_Stmt(self)


class Function(Stmt):
    def __init__(self, name: Token, params: list[Token], body: list[Stmt]):
        super().__init__()
        self.name: Final[Token] = name
        self.params: Final[list[Token]] = params
        self.body: Final[list[Stmt]] = body

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Function_Stmt(self)


class If(Stmt):
    def __init__(self, condition: Expr, thenBranch: Stmt, elseBranch: Stmt):
        super().__init__()
        self.condition: Final[Expr] = condition
        self.thenBranch: Final[Stmt] = thenBranch
        self.elseBranch: Final[Stmt] = elseBranch

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_If_Stmt(self)


class Print(Stmt):
    def __init__(self, expression: Expr):
        super().__init__()
        self.expression: Final[Expr] = expression

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Print_Stmt(self)


class Return(Stmt):
    def __init__(self, keyword: Token, value: Expr):
        super().__init__()
        self.keyword: Final[Token] = keyword
        self.value: Final[Expr] = value

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Return_Stmt(self)


class Var(Stmt):
    def __init__(self, name: Token, initializer: Expr):
        super().__init__()
        self.name: Final[Token] = name
        self.initializer: Final[Expr] = initializer

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Var_Stmt(self)


class While(Stmt):
    def __init__(self, condition: Expr, body: Stmt):
        super().__init__()
        self.condition: Final[Expr] = condition
        self.body: Final[Stmt] = body

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_While_Stmt(self)

File: golox/internal/stmt/stmt.go
package stmt

import (
	"github.com/englishlayup/crafting_interpreters/golox/internal/expr"
	"github.com/englishlayup/crafting_interpreters/golox/internal/token"
)

type Stmt[R any] interface {
	Accept(visitor StmtVisitor[R]) R
}

type StmtVisitor[R any] interface {
	VisitBlockStmt(stmt Block[R]) R
	VisitClassStmt(stmt Class[R]) R
	VisitExpressionStmt(stmt Expression[R]) R
	VisitFunctionStmt(stmt Function[R]) R
	VisitIfStmt(stmt If[R]) R
	VisitPrintStmt(stmt Print[R]) R
	VisitReturnStmt(stmt Return[R]) R
	VisitVarStmt(stmt Var[R]) R
	VisitWhileStmt(stmt While[R]) R
}

type Block[R any] struct {
	Statements []Stmt[R]
}

func (b Block[R]) Accept(visitor StmtVisitor[R]) R {
	return visitor.VisitBlockStmt(b)
}

type Class[R any] struct {
	Name       token.Token
	Superclass expr.Variable[R]
	Methods    []Function[R]
}

func (c Class[R]) Accept(visitor StmtVisitor[R]) R {
	return visitor.VisitClassStmt(c)
}

type Expression[R any] struct {
	Expression expr.Expr[R]
}

func (e Expression[R]) Accept(visitor StmtVisitor[R]) R {
	return visitor.VisitExpressionStmt(e)
}

type Function[R any] struct {
	Name   token.Token
	Params []token.Token
	Body   []Stmt[R]
}

func (f Function[R]) Accept(visitor StmtVisitor[R]) R {
	return visitor.VisitFunctionStmt(f)
}

type If[R any] struct {
	Condition  expr.Expr[R]
	Thenbranch Stmt[R]
	Elsebranch Stmt[R]
}

func (i If[R]) Accept(visitor StmtVisitor[R]) R {
	return visitor.VisitIfStmt(i)
}

type Print[R any] struct {
	Expression expr.Expr[R]
}

func (p Print[R]) Accept(visitor StmtVisitor[R]) R {
	return visitor.VisitPrintStmt(p)
}

type Return[R any] struct {
	Keyword token.Token
	Value   expr.Expr[R]
}

func (r Return[R]) Accept(visitor StmtVisitor[R]) R {
	return visitor.VisitReturnStmt(r)
}

type Var[R any] struct {
	Name        token.Token
	Initializer expr.Expr[R]
}

func (v Var[R]) Accept(visitor StmtVisitor[R]) R {
	return visitor.VisitVarStmt(v)
}

type While[R any] struct {
	Condition expr.Expr[R]
	Body      Stmt[R]
}

func (w While[R]) Accept(visitor StmtVisitor[R]) R {
	return visitor.VisitWhileStmt(w)
}
